polkadot/xcm/procedural/src/builder_pattern.rs

```diff
@@ -260,50 +263,75 @@ fn generate_builder_impl(name: &Ident, data_enum: &DataEnum) -> Result<TokenStre
-				let fields = match &variant.fields {
-					Fields::Named(fields) => {
-						let arg_names: Vec<_> =
-							fields.named.iter().map(|field| &field.ident).collect();
-						let arg_types: Vec<_> =
-							fields.named.iter().map(|field| &field.ty).collect();
-						quote! {
-							#(#docs)*
-							pub fn #method_name(self, #(#arg_names: impl Into<#arg_types>),*) -> XcmBuilder<Call, AnythingGoes> {
-								let mut new_instructions = self.instructions;
-								#(let #arg_names = #arg_names.into();)*
-								new_instructions.push(#name::<Call>::#variant_name { #(#arg_names),* });
-								XcmBuilder {
-									instructions: new_instructions,
-									state: core::marker::PhantomData,
-								}
+		.map(|variant| {
+			let maybe_builder_attr = variant.attrs.iter().find(|attr| match attr.meta {
+				Meta::List(ref list) => list.path.is_ident("builder"),
+				_ => false,
+			});
+			let builder_attr = match maybe_builder_attr {
+				Some(builder) => builder.clone(),
+				None => return Ok(None), /* It's not going to be an instruction that pays fees */
+			};
+			let Meta::List(ref list) = builder_attr.meta else { unreachable!("We checked before") };
+			let inner_ident: Ident = syn::parse2(list.tokens.clone()).map_err(|_| {
+				Error::new_spanned(
+					&builder_attr,
+					"Expected `builder(loads_holding)` or `builder(pays_fees)`",
+				)
+			})?;
+			let ident_to_match: Ident = syn::parse_quote!(pays_fees);
+			if inner_ident == ident_to_match {
+				Ok(Some(variant))
+			} else {
+				Ok(None) // Must have been `loads_holding` instead.
+			}
+		})
+		.collect::<Result<Vec<_>>>()?;
+
+	let pay_fees_methods = pay_fees_variants
+		.into_iter()
+		.flatten()
+		.map(|variant| {
+			let variant_name = &variant.ident;
+			let method_name_string = &variant_name.to_string().to_snake_case();
+			let method_name = syn::Ident::new(method_name_string, variant_name.span());
+			let docs = get_doc_comments(variant);
+			let fields = match &variant.fields {
+				Fields::Named(fields) => {
+					let arg_names: Vec<_> =
+						fields.named.iter().map(|field| &field.ident).collect();
+					let arg_types: Vec<_> =
+						fields.named.iter().map(|field| &field.ty).collect();
+					quote! {
+						#(#docs)*
+						pub fn #method_name(self, #(#arg_names: impl Into<#arg_types>),*) -> XcmBuilder<Call, AnythingGoes> {
+							let mut new_instructions = self.instructions;
+							#(let #arg_names = #arg_names.into();)*
+							new_instructions.push(#name::<Call>::#variant_name { #(#arg_names),* });
+							XcmBuilder {
+								instructions: new_instructions,
+								state: core::marker::PhantomData,
 							}
 						}
-					},
-					_ =>
-						return Err(Error::new_spanned(
-							variant,
-							"BuyExecution should have named fields",
-						)),
-				};
-				Ok(fields)
-			},
-		)?;
+					}
+				},
+				_ =>
+					return Err(Error::new_spanned(
+						variant,
+						"Both BuyExecution and PayFees have named fields",
```

Comment:
I'm not fully convinced that this `builder(pay_fees)` assumes specific instructions like `BuyExecution` or `PayFees` and their structure. Why do we assume that these instructions can't have unnamed fields?

I suggest extracting the entire block of code that parses `let load_holding_methods = load_holding_variants`, and doing the following:
- Turn it into a helper function `parse_variant_methods` (or better name).
- It can nicely parse both named and unnamed fields
- Add a parameter that specifies the result type (that is the only difference here), such as `LoadedHolding` or `AnythingGoes`.

This way, we could have a single helper function, like `parse_variant_methods`, which will make `generate_builder_impl` simpler and more deduplicated:
```
fn generate_builder_impl(..) {
    // We first require an instruction that load the holding register
    let load_holding_methods =  parse_variant_methods(data_enum, syn::parse_quote!(loads_holding), LoadedHolding);
...
    // Some operations are allowed after the holding register is loaded
    let allowed_after_load_holding_methods: ...
...
    // Then we require fees to be paid
    let pay_fees_methods =  parse_variant_methods(data_enum, syn::parse_quote!(pays_fees), AnythingGoes);
}
```

Reply:
@franciscoaguirre please do this in a followup PR, let's merge this one so we can work on top of the new fees mechanism